File: test/conformance/cpp/rows/W2.cpp
// W2.cpp: absent optional skips store (docs/FIXED-FORM-ALGORITHM.md:204-206, cell cpp/W2).
//
// Law: "An absent optional writes flag 0 and skips the payload store: a
// declared default under a clear flag is meaning too (fix 13)."
//
// Reference implementation: test/tables/fixedform_main.cpp:1401 (absent_optional_case).
//
// The table tblp3::Chain carries an optional ?Link field. When link_present is
// false, ChainFixedSave writes flag 0 and skips storing the payload, leaving
// the body buffer prefilled with zeros. The test stains caller storage for
// link with 0x5A to prove that no stained bytes reach the wire.
//
// Discriminating half (negative control): When link_present is true with the same
// 0x5A stained payload, ChainFixedSave DOES write the payload to the wire,
// proving the check discriminates between absent and present optionals.

#include "P3Table.h"
#include <cstdio>
#include <cstring>
#include <vector>

static int failures = 0;

static void check( bool ok, const char * msg )
{
    std::printf( "%s: %s\n", ok ? "PASS" : "FAIL", msg );
    if ( !ok ) { failures++; }
}

int main()
{
    // --- 1. Absent optional case: link_present = false ---
    tblp3::Chain v;
    tblp3::ChainReset( v );
    std::strcpy( v.name, "absent" );
    v.name_length = 6;
    v.link_present = false;
    v.link.value = 0x5A5A5A;
    std::memset( v.link.tag, 0x5A, sizeof( v.link.tag ) );
    v.link.tag_length = 5;

    check( (uint8_t) v.link.tag[0] == 0x5Au,
           "CONTROL: the absent payload really is stained in storage" );

    std::vector<uint8_t> w( (size_t) tblp3::ChainFixedMeasure( 1 ) );
    check( tblp3::ChainFixedSave( &v, 1, w.data(), (int64_t) w.size() ) == (int64_t) w.size(),
           "absent optional: the record saves" );

    const uint8_t * body = w.data() + tblp3::kTableFixedHeaderBytes + 4 + tblp3::ChainFixedLayoutBytes + 8;
    const size_t body_bytes = (size_t) tblp3::ChainFixedBodyBytes;

    check( std::memchr( body, 0x5A, body_bytes ) == NULL,
           "ABSENT OPTIONAL: not one byte of the absent payload reached the wire" );

    // And the reader reads what the flag says, with the payload at its defaults
    {
        tblp3::Chain back;
        tblp3::TableReport r;
        std::vector<tblp3::TableFixedEntry> plan( 1024 );
        check( tblp3::ChainFixedLoad( &back, 1, w.data(), (int64_t) w.size(), plan.data(), 1024, NULL, &r ) == 1,
               "absent optional: the record reads" );
        check( !back.link_present, "absent optional: the flag" );
        check( back.link.value == 0 && back.link.tag_length == 0,
               "absent optional: the payload reads as the wire's zeros" );
        check( r.clamped == 0 && !r.malformed && !r.refused,
               "absent optional: a clean read moves no counter" );
    }

    // --- 2. The discriminating half: the same payload, PRESENT ---
    {
        tblp3::Chain present = v;
        present.link_present = true;
        present.link.tag_length = 4;
        std::vector<uint8_t> pw( (size_t) tblp3::ChainFixedMeasure( 1 ) );
        check( tblp3::ChainFixedSave( &present, 1, pw.data(), (int64_t) pw.size() ) == (int64_t) pw.size(),
               "absent optional: the present twin saves" );
        const uint8_t * pbody = pw.data() + tblp3::kTableFixedHeaderBytes + 4 + tblp3::ChainFixedLayoutBytes + 8;
        check( std::memchr( pbody, 0x5A, (size_t) tblp3::ChainFixedBodyBytes ) != NULL,
               "NEGATIVE CONTROL: the SAME payload PRESENT really does reach the wire" );
    }

    if ( failures > 0 ) {
        std::fprintf( stderr, "RED: %d failure(s)\n", failures );
        return 1;
    }
    std::printf( "PASS: absent optional skips store\n" );
    return 0;
}

File: test/conformance/rust/rows/W2.rs
// W2.rs — rust/W2: absent optional skips store
//
// Law: docs/FIXED-FORM-ALGORITHM.md:205
//   "An absent optional writes flag 0 and skips the payload store"
//
// Schema: test/tables/P3.schema — Chain has `?Link`, a by-value optional.
// The writer zero-fills the body template, then writes only the live fields.
// When link_present is false, the Link payload stays as the template's zeros —
// not the caller's storage, not the field's declared default.

#[test]
fn test_row_w2() {
    use tblp3::*;

    // --- ABSENT: link_present = false, payload STAINED with 0x5A ---
    let mut v = ChainRow::default();
    v.name[..6].copy_from_slice(b"absent");
    v.name_length = 6;
    v.link_present = false;
    // stain the payload — a byte no clean record carries
    v.link.value = 0x5A5A5A;
    v.link.tag.fill(0x5A);
    v.link.tag_length = 5;

    // the stain is really there in storage
    assert!(
        v.link.tag[0] == 0x5A,
        "CONTROL: the absent payload is stained in storage"
    );

    // save
    let need = chain_fixed_measure(1);
    let mut w = vec![0u8; need];
    let n = chain_fixed_save(&[v], &mut w).expect("save succeeds");
    assert_eq!(n, need, "absent optional: the record saves");

    // the body starts after the header + layout length + layout block
    let body_at = TABLE_FIXED_HEADER_BYTES + 4 + CHAIN_FIXED_BLOCK.len() + 8;
    let body = &w[body_at..body_at + CHAIN_FIXED_BODY_BYTES];

    // THE LAW: not one byte of the absent payload reached the wire
    assert!(
        !body.contains(&0x5A),
        "ABSENT OPTIONAL: not one byte of the absent payload reached the wire"
    );

    // the flag on the wire is 0
    assert_eq!(
        body[20], 0,
        "absent optional: the flag is 0 on the wire"
    );

    // --- READ BACK ---
    let mut back = [ChainRow::default(); 1];
    let mut plan = [TableFixedEntry::default(); 1024];
    let mut remap = [0u16; 1024];
    let mut report = TableFixedReport::default();
    let rn = chain_fixed_load(&mut back, &w, &mut plan, &mut remap, &mut report)
        .expect("read succeeds");
    assert_eq!(rn, 1, "absent optional: the record reads");
    assert!(!back[0].link_present, "absent optional: the flag reads false");
    assert_eq!(
        back[0].link.value, 0,
        "absent optional: the payload value reads as the wire's zeros"
    );
    assert_eq!(
        back[0].link.tag_length, 0,
        "absent optional: the payload tag_length reads as the wire's zeros"
    );
    assert!(
        report.clamped == 0 && !report.malformed && !report.refused,
        "absent optional: a clean read moves no counter"
    );

    // --- NEGATIVE CONTROL: the SAME payload PRESENT ---
    let mut present = v;
    present.link_present = true;
    present.link.tag_length = 4;
    let mut pw = vec![0u8; need];
    chain_fixed_save(&[present], &mut pw).expect("present twin saves");
    let pbody = &pw[body_at..body_at + CHAIN_FIXED_BODY_BYTES];
    assert!(
        pbody.contains(&0x5A),
        "NEGATIVE CONTROL: the SAME payload PRESENT really does reach the wire"
    );
}
